hbase-client/src/main/java/org/apache/hadoop/hbase/client/MasterRegistry.java

```diff
@@ -0,0 +1,364 @@
+  // abstract out the logic needed to hedge the requests to multiple masters. For more details, look
+  // at doRPCs().
+  @VisibleForTesting
+  @FunctionalInterface
+  public interface RpcCall<RESP> {
+    RESP doRpc(ClientMetaService.BlockingInterface stub) throws ServiceException;
+  }
+
+  /**
+   * A shared RPC context between a batch of hedged RPCs. Tracks the state and helpers needed to
+   * synchronize on multiple RPCs to different masters fetching the result. All the methods are
+   * thread-safe.
+   * @param <RESP> Return response type for the RPCs.
+   */
+  private class BatchRpcCtx<RESP> {
+    // Result set by the thread finishing first. Set only once.
+    private AtomicReference<RESP> result;
+    // Caller waits on this latch being set.
+    private CountDownLatch resultsReady;
+    // Book-keeping for number of failed RPCs.
+    private AtomicInteger failedRPCs;
+
+    BatchRpcCtx() {
+      result = new AtomicReference<>();
+      // We set this to 1, so that the first successful RPC result is returned to the client.
+      resultsReady = new CountDownLatch(1);
+      failedRPCs = new AtomicInteger(0);
+    }
+
+    /**
+     * Sets the result only if it is not already set by another thread. Thread that successfully
+     * sets the result also count downs the latch.
+     * @param result Result to be set.
+     */
+    public void setResultIfNotSet(RESP result) {
+      if (this.result.compareAndSet(null, result)) {
+        resultsReady.countDown();
+      }
+    }
+
+    /**
+     * Caller can use this method to wait for results to be fetched.
+     * @param timeoutNs Waits until this timeout hits or the results are set. Whatever happens
+     *                  first.
+     * @return True if the results are ready. False otherwise.
+     */
+    public boolean waitForResults(int timeoutNs) {
+      try {
+        return resultsReady.await(timeoutNs, TimeUnit.NANOSECONDS);
+      } catch (InterruptedException e) {
+        LOG.warn("Interrupted while waiting for batched master RPC results. Aborting wait.", e);
+        return false;
+      }
+    }
+
+    /**
+     * Helper to increment the number of failed RPCs.
+     */
+    public void incrementFailedRPCs() {
+      failedRPCs.incrementAndGet();
+    }
+
+    /**
+     * Onus is on the caller to wait for the results and call this.
+     * @return the current result.
+     */
+    public RESP getResult() {
+      return result.get();
+    }
+  }
+
+  /**
+   * A runnable implementation of an RPC call to a given master. Updates the results in a shared
+   * rpc context.
+   * @param <RESP> Response type of the RPC.
+   */
+  private class MasterRpc<RESP> implements Runnable {
+    private final BatchRpcCtx<RESP> rpcCtx;
+    private final ServerName master;
+    private final RpcCall<RESP> rpcCall;
+    private final Function<RESP, Boolean> isValidResp;
+    private final String debugStr;
+
+    MasterRpc(BatchRpcCtx<RESP> rpcCtx, ServerName master, RpcCall<RESP> rpcCall,
```

Comment:
We have a whole world of Callable+Call and it seems like this is introducing a new one of RpcCall... We need this new one? (I don't know.... just asking... because it gets hard tracking Callable/Call/Caller... and adding variants makes for more to decipher... FYI).

Reply:
We don't need a new one, right? Use existing RPC/Call facilities and set a retry policy using the factory where there are no retries.